src/persist-client/src/write.rs

```diff
@@ -266,73 +210,225 @@ where
-        let value = match Self::encode_batch(&desc, updates) {
-            Ok(x) => x,
-            Err(err) => return Ok(Err(err)),
-        };
-        let keys = if let Some(value) = value {
-            let key = Uuid::new_v4().to_string();
-            let () = retry_external("compare_and_append::set", || async {
-                // If MultiBlob::set took value as a ref, then we wouldn't have
-                // to clone here.
-                self.blob
-                    .set(
-                        Instant::now() + FOREVER,
-                        &key,
-                        value.clone(),
-                        Atomicity::RequireAtomic,
-                    )
-                    .await
-            })
-            .await;
-            vec![key]
-        } else {
-            vec![]
+        let mut batch = match self
+            .batch(updates, expected_upper.clone(), new_upper.clone())
+            .await
+        {
+            Ok(batch) => batch,
+            Err(invalid_usage) => return Ok(Err(invalid_usage)),
         };
 
+        match self
+            .compare_and_append_batch(&mut batch, expected_upper, new_upper)
+            .await
+        {
+            ok @ Ok(Ok(Ok(()))) => ok,
+            err @ _ => {
+                // We cannot delete the batch in compare_and_append_batch()
+                // because the caller owns the batch and might want to retry
+                // with a different `expected_upper`. In this function, we
+                // control the batch, so we have to delete it.
+                batch.delete().await?;
+                err
+            }
+        }
+    }
+
+    /// Appends the batch of updates to the shard and downgrades this handle's
+    /// upper to `upper`.
+    ///
+    /// The innermost `Result` is `Ok` if the updates were successfully written.
+    /// If not, an `Upper` err containing the current writer upper is returned.
+    /// If that happens, we also update our local `upper` to match the current
+    /// upper. This is useful in cases where a timeout happens in between a
+    /// successful write and returning that to the client.
+    ///
+    /// In contrast to [Self::compare_and_append_batch], multiple [WriteHandle]s
+    /// may be used concurrently to write to the same shard, but in this case,
+    /// the data being written must be identical (in the sense of
+    /// "definite"-ness). It's intended for replicated use by source ingestion,
+    /// sinks, etc.
+    ///
+    /// A `upper` of the empty antichain "finishes" this shard, promising that
+    /// no more data is ever incoming.
+    ///
+    /// The batch may be empty, which allows for downgrading `upper` to
+    /// communicate progress. It is possible to heartbeat a writer lease by
+    /// calling this with `upper` equal to `self.upper()` and an empty `updates`
+    /// (making the call a no-op).
+    ///
+    /// IMPORTANT: In case of an erroneous result the caller is responsible for
```

Comment:
this is definitely true for `compare_and_append_batch` but is it also true for `append_batch`? I think maybe not but I don't want to be prescriptive about it yet. keep this as-is, but I wanted to get you thinking about it

Reply:
I agree the contract is simple, but making the caller deal with the batch is not simple for them. If the contract of this method can be "we always use/delete the batch" that's simpler for them :)

Reply:
True! :sweat_smile: 

I could change `append_batch` to take a `self` and keep `compare_and_append_batch` on `&mut self` to make that distinction clear. Our internal logic would be slightly more complicated in `append_batch` but that should be worth it.